Kernel/Devices/SerialDevice.cpp

```diff
@@ -140,4 +92,212 @@ u8 SerialDevice::get_line_status() const
     return m_base_addr.offset(5).in<u8>();
 }
 
+ssize_t SerialDevice::on_tty_write(const UserOrKernelBuffer& buffer, ssize_t size)
+{
+    if (!size)
+        return 0;
+
+    if (!(get_line_status() & EmptyTransmitterHoldingRegister))
+        return 0;
+
+    auto result = buffer.read_buffered<128>(size, [&](u8 const* data, size_t data_size) {
+        for (size_t i = 0; i < data_size; i++)
+            put_char(data[i]);
+        return data_size;
+    });
+    if (result.is_error())
+        return result.error();
+    return (ssize_t)result.value();
+}
+
+void SerialDevice::echo(u8 ch)
+{
+    put_char(ch);
+}
+
+void SerialDevice::discard_pending_input()
+{
+    set_fifo_control(ClearReceiveFIFO);
+}
+
+void SerialDevice::discard_pending_output()
+{
+    set_fifo_control(ClearTransmitFIFO);
+}
+
+void SerialDevice::wait_until_pending_output_completes()
+{
+    while ((get_line_status() & EmptyDataHoldingRegisters) == 0)
+        ;
+}
+
+int SerialDevice::change_baud(speed_t in_baud, speed_t out_baud)
+{
+    if (in_baud != out_baud) {
+        dbgln("{}: Input and output speed must be the same", m_tty_name);
+        return -ENOTSUP;
+    }
+    auto maybe_baud = serial_baud_from_termios(in_baud);
+    if (maybe_baud.has_value()) {
+        set_baud(maybe_baud.value());
+    } else {
+        dbgln("{}: Tried to set unsupported baud speed", m_tty_name);
+        return -ENOTSUP;
+    }
+    return 0;
+}
+
+Optional<speed_t> SerialDevice::termios_baud_from_serial(Baud baud)
+{
+    switch (baud) {
+    case Baud::Baud50:
+        return B50;
+    case Baud::Baud110:
+        return B110;
+    case Baud::Baud300:
+        return B300;
+    case Baud::Baud600:
+        return B600;
+    case Baud::Baud1200:
+        return B1200;
+    case Baud::Baud2400:
+        return B2400;
+    case Baud::Baud4800:
+        return B4800;
+    case Baud::Baud9600:
+        return B9600;
+    case Baud::Baud19200:
+        return B19200;
+    case Baud::Baud38400:
+        return B38400;
+    case Baud::Baud57600:
+        return B57600;
+    case Baud::Baud115200:
+        return B115200;
+    default:
+        return {};
```

Comment:
instead of returning an optional and handling this in each call site, how about just VERIFYING_NOT_REACHED here? (this will also make it more clear for the next person who changes the code and causes a mismatch between serial/termios BAUDs)

Reply:
Actually there's a mismatch now. I'll add the missing baud then.